### ISSUE TITLE
cVM powered off, but container appears as running during ps -a

### ISSUE THREAD

**Issue by malikkal** [vmware/vic#4872]
engine version: v1.1.0-9852-e974a51

Steps to reproduce.

`docker -H ip:port --tls run -it --cpuset-cpus 2 --memory 4g --name <name> --net <network> registry/project/hello-world`

Ctrl-c to exit the shell, which will power off the cVM, followed by a docker ps will display the container as running. docker rm or rm -f will fail with:

`You cannot remove a running container. Stop the container before attempting removal or use -f`

delete the cVM from vSphere. I am aware, this is not recommended; followed by docker ps will display them as **still** running and a docker stop will result in:

`No such container: ID`

power off the vAPP and power on will result in correct output for docker ps. however, redeploying the same image will result in.

`Server error from portlayer: Invalid configuration for device '1'.`

**Comment by malikkal** [comment]
...pardon me. I couldn't upload the logs here, as it was not tested in the dev environment.  I have created a directory github4872 in the ftpsite and uploaded the logs there.

**Comment by hickeng** [comment]
@malikkal The latter part of the problem (invalid configuration for device '1') is #2928. The incorrect state reporting requires investigation.

@dbarkelew Is there a KB in-progress for the out-of-band delete? I know @chengwang86 wrote the docs from the engineering side, but don't know where it stands in the GSS/publication path.
Could you also move the ftpsite files into a PR so dev can access them.

**Comment by hickeng** [comment]
Estimate is for initial investigation of the container state aspect - I expect the estimate to be updated once more knowledge is available.

**Comment by sflxn** [comment]
@malikkal Can I get some more info?  This hello-world container is your own or docker's?  I ask because the hello-world container I am familiar with only dumps out a greeting and some info and takes you back to your original shell so -it will have no effect since it doesn't put you into another shell.  I must assume you're running another container.  Can you provide me some info on how to get the one you're using?

I also tried this with just busybox, which puts me into sh, but ctrl-c cannot break me out of the container's sh shell.

**Comment by hickeng** [comment]
@malikkal Please can you open an SR with GSS to get us the logs. It turns out that there's automation hooked into that process that makes logs uploaded to ftpsite available internally, I assume to enforce information access controls around customer confidential logs. I've not found a way to access those files directly from the ftpsite.
Github is our day-to-day task tracking, so issues opened here will be seen and addressed, but GSS is where the SLAs and so on are applied for paid support. Once we've got an issue open with GSS and reflected in github then we have a channel for passing logs back and forth securely and am happy to continue with github based interaction, although I suspect that the same SLA statement applies.

My apologies for the delay and collision of processes.

**Comment by chengwang86** [comment]
This might relate to https://github.com/vmware/vic/issues/3434

**Comment by chengwang86** [comment]
I kind of reproduced the former part of the error by performing the following steps on a VCH deployed on a vCenter:
1. `docker run -it ubuntu` then `ctrl-c` and `exit`
2. `docker ps -a` immediately after 1
3. `docker ps -a` again after long enough time

Then output from 2 is:
```
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                     PORTS               NAMES
e7fc799de532        ubuntu              "/bin/bash"         3 minutes ago       Up 3 minutes                                   sleepy_swirles
94706a0240b1        hello-world         "/hello"            8 minutes ago       Exited (0) 7 minutes ago                       wonderful_tesla
```

The output from 3 is:
```
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                        PORTS               NAMES
e7fc799de532        ubuntu              "/bin/bash"         3 minutes ago       Exited (130) 18 seconds ago                       sleepy_swirles
94706a0240b1        hello-world         "/hello"            8 minutes ago       Exited (0) 7 minutes ago                          wonderful_tesla
```

Logs can be found here 
[container-logs (1).zip](https://github.com/vmware/vic/files/1002560/container-logs.1.zip)

**Comment by sflxn** [comment]
Still waiting for logs specific to the original poster's scenario.  However, this may be fixed by this PR:

https://github.com/vmware/vic/pull/5047

**Comment by mhagen-vmware** [comment]
closing for now, @malikkal if you see this again please re-open and provide the logs.  Thanks!

**Comment by malikkal** [comment]
Sorry, I was a bit late here. But, I have provided 3 sets of VCH logs with similar issues via the support case. Any help in fixing this on priority is highly appreciated. Thanks.

**Comment by cgtexmex** [comment]
@malikkal thanks for the logs -- I have them and am looking into the issues..

**Comment by sflxn** [comment]
After looking at the logs, @caglar10ur and I saw this in the tether logs:

```
May 18 2017 13:51:11.705+08:00 INFO  Launching process for session eb0ce15a8e749c77c4834621095088105cb08260b7dc0b3462c1a18ecfd79725
May 18 2017 13:51:11.705+08:00 INFO  Trying to establish a connection
May 18 2017 13:51:11.708+08:00 INFO  opening ttyS0 for backchannel
May 18 2017 13:51:11.709+08:00 INFO  setting terminal to raw mode
May 18 2017 13:51:11.709+08:00 INFO  creating raw connection from ttyS0 (fd=7)
May 19 2017 11:24:44.750+08:00 INFO  Ready to service attach requests
May 19 2017 11:24:44.790+08:00 INFO  received global request type container-ids
[77282.819779] EXT4-fs error (device sda): ext4_find_entry:1450: inode #2: comm tether: reading directory lblock 0
May 19 2017 11:24:44.835+08:00 ERROR failed to start container process: fork/exec bin/bash: no such file or directory
May 19 2017 11:24:44.883+08:00 ERROR failed to launch bin/bash for eb0ce15a8e749c77c4834621095088105cb08260b7dc0b3462c1a18ecfd79725: failed to start container process: fork/exec bin/bash: no such file or directory
May 19 2017 11:24:44.885+08:00 ERROR failed to launch bin/bash for eb0ce15a8e749c77c4834621095088105cb08260b7dc0b3462c1a18ecfd79725: failed to start container process: fork/exec bin/bash: no such file or directory
May 19 2017 11:24:44.887+08:00 INFO  Stopping extension Attach
May 19 2017 11:24:44.888+08:00 INFO  Stopping extension Toolbox
[77282.879953] reboot: Power down
```

It's not missing events as we originally thought.  The container VM didn't start.  The logs seem to indicate some kind of disk issue.

**Comment by hickeng** [comment]
@malikkal #5343 was the problems you hit with 2 of the 3 problem VCHs - I do not know why you were losing the connection to VC however. There's three possibilities I can think of:
1. network disruption of some kind
2. direct deleting of sessions (whether via govc session.rm, UI or otherwise)
3. unknown interference 

If you are aware of (1) or (2) occurring then #5350 adds the necessary resiliency. If neither is the case we still have an unidentified cause. I will assume (3) for now and continue to dig through logs to see if there's a visible root cause.

I will look further into the other VCH diagnostics today - this will encompass the datum that @sflxn provided in his last comment.

**Comment by malikkal** [comment]
@hickeng, Thanks much for your help.

I am quite confident that we wouldn't be deleting those sessions manually or via govmomi etc.. A quick check shows that we have established sessions since 12th May. However, that coincided with our 6.0 u3b upgrade of our vCenter. May I know how often you see losing the connections, please? That would prompt me to think if its network related or something else...

BTW, the upgrade failed for the fix provided; please refer #5343

**Comment by hickeng** [comment]
@malikkal Looks like the connection was dropped on May 12th - that's the only drop I can see - so seems likely to relate to the vCenter upgrade:
May 12 2017 06:08:46.477Z WARN  session keepalive error: ServerFaultCode: The session is not authenticated.
May 12 2017 06:08:46.539Z INFO  session keepalive re-authenticated

Will look into the upgrade failure.

**Comment by malikkal** [comment]
Today, one of the VCH (vch02a - a.b.x.35), where I had submitted logs earlier broke fully; i.e. no ps or ps -a and no info. I am uploading the logs via support case now. Thanks.

**Comment by malikkal** [comment]
build 11278 didn't help, except the upgrade went through. In summary,

1. As above, it didn't solve the accuracy of the ps output; i.e. powered off cVMs still not displayed.
2. it created a new issue. I couldn't ping IPs of the cVMs.
`From IP icmp_seq=1 Destination Port Unreachable`
3. docker network inspect <<network>> does show its receiving an IP though.

![image](https://user-images.githubusercontent.com/16984768/26866563-35862b96-4b94-11e7-98d8-e69ee933a00c.png)

**Comment by hickeng** [comment]
@malikkal The 11222 build (and later) fixes the issue of a dropped VC connection impacting state updates. It does not address the problem showing up in the VCH in your recent comments; that is demonstrating a separate issue.
I believe that if you were to delete vigilant_sammet and restart the VCH you'd be back to good function. More details below:

As best I can determine there was a modification to the configuration for `vigilant_sammet` while the VM was powered on which has left it with invalid configuration data. This can be seen from the following entries in the `vmware.log`:
```
2017-05-24T02:24:51.494Z| vmx| I125: VigorTransportProcessClientPayload: opID=599785cc-7e-e4f3 seq=2009: Receiving ConfigParams.SetVolatileEntry request.
2017-05-24T02:24:51.494Z| vmx| A100: ConfigDB: Unsetting "guestinfo.vice./version/PluginVersion"
2017-05-24T02:24:51.509Z| vmx| I125: VigorTransport_ServerSendResponse opID=599785cc-7e-e4f3 seq=2009: Completed ConfigParams request.
```
where the only updates we **should** be seeing while the VM is powered on would look like:
```
2017-05-24T02:31:22.660Z| vmx| I125: VigorTransportProcessClientPayload: opID=37dc4ce8-bb-ecc0 seq=2716: Receiving ConfigParams.SetVolatileEntry request.
2017-05-24T02:31:22.660Z| vmx| A100: ConfigDB: Unsetting "common/name"
2017-05-24T02:31:22.660Z| vmx| A100: ConfigDB: Setting common/name = "vigilant_sammet"
2017-05-24T02:31:22.690Z| vmx| I125: VigorTransport_ServerSendResponse opID=37dc4ce8-bb-ecc0 seq=2716: Completed ConfigParams request.
```

Precisely how this occurred is what I'm chasing down now - there is a vSphere test/set mechanism (ChangeVersion) that we use to avoid this scenario but we've evidently got an edge case that's unhandled.

The problem you're currently observing with `ps` not returning anything looks to be a secondary failure on the error path created by the invalid configuration, but I've not yet confirmed that hypothesis.

**Comment by malikkal** [comment]
@hickeng, thanks. I will delete the problem cVM and observe. Could you please try run --rm that should auto kill the VM on exit? I remember this is how the dev folks are running and on the last attempt the VCH went unresponsive before the upgrade.

But, may I know what about the second issue introduced by the new build? Any cVM spawned in the new build cannot be pinged. I haven't uploaded the logs yet for this.

**Comment by hickeng** [comment]
@malikkal Drat - the ping is my fault for not mentioning it. We've added iptables configuration for containers that are attached to external networks in preparation for NFS volumes; we don't want people unexpectedly exposing container ports to an open network just because they were wanting to consume an NFS volume without NAT overhead and therefore added a container network.
This means that when running a container on a container network you need to declare the ports accessible in the same manner as when forwarding via NAT - the idea here being to preserve the whitelist by default behaviour of the NAT access:

Without exposing port:
```
vagrant@devbox:~/vicsmb/src/github.com/vmware/vic$ docker run -it --net=external tomcat:alpine
vagrant@devbox:~/vicsmb/src/github.com/vmware/vic$ nmap -p 8080 192.168.78.133

Starting Nmap 7.01 ( https://nmap.org ) at 2017-06-07 21:46 UTC
Nmap scan report for 192.168.78.133
Host is up (0.00093s latency).
PORT     STATE  SERVICE
8080/tcp closed http-proxy

Nmap done: 1 IP address (1 host up) scanned in 0.11 seconds
vagrant@devbox:~/vicsmb/src/github.com/vmware/vic$
```

With exposed port 8080 is reported as open:
```
vagrant@devbox:~/vic$ docker run -dit --net=external -p 8080 tomcat:alpine
vagrant@devbox:~/vicsmb/src/github.com/vmware/vic$ nmap -p 8080 192.168.78.132

Starting Nmap 7.01 ( https://nmap.org ) at 2017-06-07 21:45 UTC
Nmap scan report for 192.168.78.132
Host is up (0.0017s latency).
PORT     STATE SERVICE
8080/tcp open  http-proxy

Nmap done: 1 IP address (1 host up) scanned in 0.11 seconds
vagrant@devbox:~/vicsmb/src/github.com/vmware/vic$
```

The default drop all includes ICMP. For the moment we've had to include the iptables binary in the base image to do this until we can implement via IPTC or similar. As such ping can be enabled via:
```
LD_LIBRARY_PATH=/.tether/lib:/.tether/lib64 /.tether/lib64/ld-linux-x86-64.so.2 ./iptables -I INPUT 1 -p icmp --icmp-type 8 -s 0/0 -i eth0 -m state --state NEW,ESTABLISHED,RELATED -j ACCEPT
```
which should result in an `iptables -L` output similar to the following (has 8080 exposed for tomcat):
```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
ACCEPT     all  --  anywhere             anywhere             state ESTABLISHED
ACCEPT     icmp --  anywhere             anywhere             icmp echo-request state NEW,RELATED,ESTABLISHED
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:8080
REJECT     all  --  anywhere             anywhere             reject-with icmp-port-unreachable

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
```

Of note - the iptables config work is impacted by #3000 so we don't yet support `-P` to expose all ports declared via `EXPOSE`. 
We are aiming to support port forwarding in this scenario as well (so tomcat can be accessed on `80` and not the hardcoded application port `8080`) however that's not implemented yet.

Do you need ping enabled by default? I've no problem at all with adding ping to the default firewall set.

**Comment by malikkal** [comment]
Thanks @hickeng.

> This means that when running a container on a container network you need to declare the ports accessible in the same manner as when forwarding via NAT - the idea here being to preserve the whitelist by default behaviour of the NAT access:

IMHO, this is bad. Is it possible to disable this during startup for cVMs that use container networks; i.e. turn off iptables completely? Either we should expose all (as an option) while using container networks or give option to disable it.

Undeterred access (without NAT and iptables overhead) while **using a container network** is actually one of the selling points to go ahead with VIC.

For example, in grid computing use cases, they dynamically open a lot of ports and there's bilateral communication between head and compute nodes. I am not sure, how this will impact otherwise.

**Comment by malikkal** [comment]
>  I will delete the problem cVM and observe. 
I did this. so far, no issues.

Also, I restarted the VC. Interestingly, even after the sessions were broken for the older VCH build, I could continue to do the normal activities on all the VCH instances; i.e. info, ps/ps-a and deployment of cVMs after VC was up.

**Comment by hickeng** [comment]
@malikkal we're intentionally biasing towards secure-by-default configuration hence the default firewall behaviour. We can definitely add configuration mechanism to address your use case. Approaches that come immediately to mind:
1. mark the container-network as a whole with `vic-machine`, e.g. `--container-network-firewall=<name>:[open|outbound|exposed|published|blocked]`, or
2. per container with a similar set of arg/values

In this case the options would have the following behaviour (from most permissive to least):
* open - no firewall
* exposed - ports defined via `Dockerfile:EXPOSE` or `--expose` will be open for inbound
* published - ports explicitly published via `-p`  or `-P` will be open for inbound
* outbound - outbound connections only (implied for open, exposed, and published)
* blocked - firewall blocks everything by default and the container process manages access via iptables

While we could provide arbitrary granularity I don't think it adds much value for the increase in usage complexity at this time, and comes with a significant development overhead.
It's worth noting that you can EXPOSE a range of ports, so it's quite possible to allow for ftp or similar dynamic incoming port allocation so long as there's sufficient knowledge to express the range.

Would the `--container-network-firewall` suffice for you?

**Comment by malikkal** [comment]
@hickeng Many thanks much for considering. Yes, please, so long as we have an option to be permissive for container-network it should be okay. BTW, I am curious, since this is affecting the cVM, the firewall rules will be dynamically pushed on startup?

While, we are waiting for this to be enabled, I am testing the new build 11278 with an image, where iptables will be stopped on start-up...

**Comment by malikkal** [comment]
**Important observation**

> In addition, I restarted the vCenter. Interestingly, even after the sessions were broken for the older VCH build, I could continue to do the normal activities on all the VCH instances; i.e. info, ps/ps-a and deployment of cVMs after VC was up.

Spoke too early.

1. After restart of vCenter, everything is not okay for earlier builds.
a. couldn't `rm -f` running containers.
b. couldn't `stop`, but `restart` worked followed by `rm -f`.
c. `attach` failed for running containers.
d. this eventually deteriorated and `restart` also start failing leaving the container in `starting` mode...

2. I **rebooted** the older build appliance.
a. this partially restored some functionality like `rm -f`
b. however, `ps` output was inconsistent, couldn't pickup up powered off containers for example.
c. after the new container was deployed, even though it was up and attached to; `ps` will report this as `starting`, instead of `up`.

I will upload two set of logs to the case, if it would help; one before restart of appliance and one after restart of appliance.

The **good news** is; I didn't observe these issues in the newer build yet and dev folks have already started working on the new VCH, with an image tweaked to stop iptables (in the interim).

**Comment by hickeng** [comment]
@malikkal There's a couple of different issues that you've hit. There's one relating to the event stream not correctly re-establishing after the VC connection goes down. That's addressed in the build you're currently working with. Restarting the old VCH should have the same effect.

I believe - and am still working to recreate and confirm - that the powered off containers not presenting and getting stuck in starting is due to a missed secondary state check in a commit path that would protect against a race. What I can confirm is that the configuration embedded in `vigilant_sammet` from the earlier logs was damaged - I presume at this time that the `Starting` state is because the container process isn't launching as expected because of that damage. The tether.debug for the stuck cVM may shed light on it's actual initialization state.

Something in your env is triggering the timing needed to see this issue - my focus at the moment is to attempt to recreate that timing here so I can confirm my fix that's based on code inspection.

**Comment by hickeng** [comment]
Merged a set of changes in #5532 that address a number of the issues described in this issue.

I believe #5532 addresses the behavioural issues described here, and #5641 will address configuring the trust level of a container network once it's merged.

I'm moving this to verify - once #5641 is merged we should request @malikkal confirm function and then close if accepted.

**Comment by malikkal** [comment]
Incidentally, today, the issue is back again, with slight variations. Here are the sequence of events.

1. We had an unplanned vCenter downtime today.
2. Dev folks later complained they couldn't attach to running containers after vCenter was back online. Error message was:
```
Error: No such container: <container name>
Server error from portlayer: Unable to find a device with desired backing
```
Tried to get VCH into debug mode. But, noticed that tools aren't running and debug failed.
![image](https://user-images.githubusercontent.com/16984768/28266423-513ec3fe-6b27-11e7-98ca-978ec803092b.png)

3. Rebooted VCH and tools came back online and debug worked.
4. Noticed that VCH picked new IP from DHCP. Side note: Need to think about reservation or static IP for production.
5. VCH didnt respond to info or ps for long time. Notice from logs that its refreshing image cache and rebuilding metadata, which went on for quite long time.
6. Once this operation is completed, I could get info and ps working.
7. status now:
a. provisioning new container get stuck in starting mode and failing in command line with: context deadline exceeded.
b. remove operation is working, but extremely slow and noted: 
`Failed to enable Destroy_Task for VirtualMachine:vm-63091: ServerFaultCode: Permission to perform this operation was denied.`
c. Reaping old sessions going on for a while in vicadmin.log. Not sure if this is related.

I will upload few megabytes of logs. Thanks.

**Comment by mhagen-vmware** [comment]
Based on last comment, reopening and moving back to backlog.

**Comment by malikkal** [comment]
Some additional info, if it helps.

Tried to test the behavior after an upgrade using vic_12396.tar.gz and most of the functions worked well, though docker attach failed for existing cVMs with the same error `Unable to find a device with desired backing` .

However, I could provision new cVMs and they worked mostly well. Reverting the snapshot produced some interesting behavior, documented in #5754.

**Comment by hickeng** [comment]
@malikkal @corrieb ran some additional targeted VC restart testing here to see if we can replicate the problems you observed but we didn't manage to recreate it; will keep looking.

@caglar10ur Any idea about the attach error? Only path I can see that would result in this is if `handle.Config.Hardware.Device` is not fully populated.

**Comment by dougm** [comment]
I haven't been able to reproduce the power state issue yet, but can force the issue by intentionally missing VmPoweredOffEvents:

```diff
modified   lib/portlayer/event/collector/vsphere/collector.go
@@ -184,7 +184,6 @@ func evented(ec *EventCollector, page []types.BaseEvent) {
 		case *types.VmGuestShutdownEvent,
 			*types.VmPoweredOnEvent,
 			*types.DrsVmPoweredOnEvent,
-			*types.VmPoweredOffEvent,
 			*types.VmRemovedEvent,
 			*types.VmSuspendedEvent,
 			*types.VmMigratedEvent,
```
Should we consider adding a mechanism to periodically sync state to handle the case where we might miss this even?  cc @cgtexmex

**Comment by cgtexmex** [comment]
@dougm _possibly_ ... I've never been able to create an `event storm` that would result in missed events.  Additionally, I believe that @hickeng would view periodic state syncing as a responsibility of the govmomi gateway....so maybe better left for that implementation since we have no clear indicator that events are being missed.

**Comment by dougm** [comment]
The container VM cannot be removed via 'docker rm' or 'docker rm -f', even though it is powered off.  I'll see if we can do a power state sync in the 'rm -f' path at least for now.

**Comment by dougm** [comment]
Doesn't help the 'ps' state issue, but workaround does make it possible to rm the container VM: PR #6083

**Comment by mdubya66** [comment]
Closing this. @malikkal if it's seen again with a newer build let us know and we can re-open it.
